sharding/proposer/service.go

```diff
@@ -46,42 +60,57 @@ func (p *Proposer) Start() {
 // Stop the main loop for proposing collations.
 func (p *Proposer) Stop() error {
 	log.Info(fmt.Sprintf("Stopping proposer service in shard %d", p.shardID))
+	defer p.cancel()
+	p.txpoolSub.Unsubscribe()
 	return nil
 }
 
+// proposeCollations listens to the transaction feed and submits collations over an interval.
 func (p *Proposer) proposeCollations() {
+	requests := make(chan *types.Transaction)
+	p.txpoolSub = p.txpool.TransactionsFeed().Subscribe(requests)
+
+	for {
+		ctx, cancel := context.WithTimeout(p.ctx, 10*time.Second)
```

Comment:
Why do you need to define a new context here? Let's just use p.ctx.Done() in the select block. Also, this will cancel the goroutine right after the select block...we instead want this to run forever.

Reply:
I derived a new context here to mimic the behavior [here](https://github.com/rauljordan/go-ethereum/commit/05ce7ae677a847e1f6369c8f1f81bfaed44094e0#diff-2a673836cc5cb435fcd7be724b985d3cR50). If a single loop takes more than 10 seconds, whatever process is running gets canceled and the loop restarts. On the other hand if the parent context gets canceled we want the goroutine to be exit entirely.

Reply:
This looks good to me.